diff --git a/docroot/sites/all/modules/contrib/civicrm/CRM/Mailing/BAO/MailingJob.php b/docroot/sites/all/modules/contrib/civicrm/CRM/Mailing/BAO/MailingJob.php
index 4bd01509d..2fdd4e935 100644
--- a/docroot/sites/all/modules/contrib/civicrm/CRM/Mailing/BAO/MailingJob.php
+++ b/docroot/sites/all/modules/contrib/civicrm/CRM/Mailing/BAO/MailingJob.php
@@ -39,7 +39,7 @@ require_once 'Mail.php';
  * Class CRM_Mailing_BAO_MailingJob
  */
 class CRM_Mailing_BAO_MailingJob extends CRM_Mailing_DAO_MailingJob {
-  CONST MAX_CONTACTS_TO_PROCESS = 8000;
+  CONST MAX_CONTACTS_TO_PROCESS = 5000;
 
   /**
    * class constructor
@@ -508,8 +508,7 @@ VALUES (%1, %2, %3, %4, %5, %6, %7)
                             ON  $eqTable.id = $ebTable.event_queue_id
                     WHERE       $eqTable.job_id = " . $this->id . "
                         AND     $edTable.id IS null
-                        AND     $ebTable.id IS null
-                        AND    $contactTable.is_opt_out = 0";
+                        AND     $ebTable.id IS null";
 
     if ($mailing->sms_provider_id) {
       $query = "
@@ -529,9 +528,19 @@ VALUES (%1, %2, %3, %4, %5, %6, %7)
                             ON  $eqTable.id = $ebTable.event_queue_id
                     WHERE       $eqTable.job_id = " . $this->id . "
                         AND     $edTable.id IS null
-                        AND     $ebTable.id IS null
-                        AND    $contactTable.is_opt_out = 0";
+                        AND     $ebTable.id IS null";
     }
+    
+    $isaReminderMailing = FALSE;
+    if (function_exists('trackmailing_is_tracking_mailing')) {
+      // track mailing extension controls if to send a mailing based on bulk flag. 
+      // lets give that control to extension.
+      $isaReminderMailing = trackmailing_is_tracking_mailing($mailing->id);
+    }
+    if (!$isaReminderMailing) {
+      $query .= " AND    $contactTable.is_opt_out = 0";
+    }
+    
     $eq->query($query);
 
     static $config = NULL;
